src/track/beats.cpp

```diff
@@ -661,6 +715,210 @@ std::optional<BeatsPointer> Beats::tryScale(BpmScale scale) const {
             m_subVersion));
 }
 
+std::optional<BeatsPointer> Beats::tryAdjustTempo(
+        audio::FramePos position, TempoAdjustment adjustment) const {
+    auto markers = m_markers;
+
+    // Retrieve the Beat Marker before the current position.
+    //
+    // Hence, we first get the marker *after* the position, then decrement the
+    // iterator.
+    // If there is no marker before the current position, the marker after the
+    // position can be used because the beats before the first marker are
+    // interpolated using the same beat length as the beats after the first
+    // marker.
+    auto markerIt = std::upper_bound(markers.begin(),
+            markers.end(),
+            position,
+            isFramePosLessThanBeatMarker);
+    if (markerIt != markers.begin() && position < m_lastMarkerPosition) {
+        markerIt--;
+    }
+
+    // At this point, `markerIt` points to the beat marker that we need to
+    // modify (or `cend()`, which means that we need to modify the BPM of the
+    // last marker).
+    Bpm lastMarkerBpm = m_lastMarkerBpm;
+    if (markerIt == markers.end()) {
+        lastMarkerBpm += (adjustment == TempoAdjustment::Faster) ? kBpmAdjustStep : -kBpmAdjustStep;
+        if (!lastMarkerBpm.isValid()) {
+            qWarning() << "Beats: Tempo adjustment would result in invalid Bpm!";
+            return std::nullopt;
+        }
+    } else {
+        const auto marker = *markerIt;
+        const int adjustedBeatsTillNextMarker = marker.beatsTillNextMarker() +
```

Comment:
I think this requires a comment. 
Is this correct, that you adjust the tempo only by removing or adding a full measure? 
In case we are sure the following beat is at a downbeat that will work. 
However I am afraid in many cases there is somewhere a tempo or rhythm change and the user will become crazy because of the big step width not solving the issue. 
Will it still possible to fine tune the tempo? 
For my feeling the fine tuning whit shifting the rest shall become default.

Reply:
> Is this correct, that you adjust the tempo only by removing or adding a full measure?

Yes, this is intentional.

> However I am afraid in many cases there is somewhere a tempo or rhythm change and the user will become crazy because of the big step width not solving the issue.

I don't understand. Let's say you have a track with multiple tempos. Say you marked a downbeat near the beginning of the track and a downbeat near the end. Maybe the tempo between those two markers is incorrect. Whatever happens, changing the tempo between the two markers should never move any marker, because the downbeats at the two markers are already marked correctly and won't move, regardless of the tempo.

> Will it still possible to fine tune the tempo?

Sure, if may move one of the markers by moving the play cursor near them and pressing *set marker* (this will move them if they are close).

Reply:
Most is related to my comment above: https://github.com/mixxxdj/mixxx/pull/4489#discussion_r750285768


I am a bit afraid of the fore and back scrolling, this idea involves. 
But it could work.

Especially if we snap to onsets ...
Another topic that need to be well considered.